# 2. UCI-SECOM(결측값 900개 제거/오버샘플링 X/SVM)

### 📌 데이터 분석 과정
- 데이터 전처리 : Null 값이 900개 넘는 칼럼 삭제하고 남는 NaN값은 0으로 대체
- 스케일링: StandardScaler, PCA
- 모델링 : SVM
- 교차 검증 : GridsearchCV
- 평가 : Confusion matrix, Accuracy score, Recall score, F1_score, Precision score

In [1]:
#Import Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings(action='ignore')

In [2]:
# Read the data
data = pd.read_csv('uci-secom.csv')

In [3]:
# Data shape 및 상위 5개 열 보기
print(data.shape)
data.head()

(1567, 592)


Time        0        1          2          3       4      5  \
0  2008-07-19 11:55:00  3030.93  2564.00  2187.7333  1411.1265  1.3602  100.0   
1  2008-07-19 12:32:00  3095.78  2465.14  2230.4222  1463.6606  0.8294  100.0   
2  2008-07-19 13:17:00  2932.61  2559.94  2186.4111  1698.0172  1.5102  100.0   
3  2008-07-19 14:43:00  2988.72  2479.90  2199.0333   909.7926  1.3204  100.0   
4  2008-07-19 15:22:00  3032.24  2502.87  2233.3667  1326.5200  1.5334  100.0   

          6       7       8  ...       581     582     583     584      585  \
0   97.6133  0.1242  1.5005  ...       NaN  0.5005  0.0118  0.0035   2.3630   
1  102.3433  0.1247  1.4966  ...  208.2045  0.5019  0.0223  0.0055   4.4447   
2   95.4878  0.1241  1.4436  ...   82.8602  0.4958  0.0157  0.0039   3.1745   
3  104.2367  0.1217  1.4882  ...   73.8432  0.4990  0.0103  0.0025   2.0544   
4  100.3967  0.1235  1.5031  ...       NaN  0.4800  0.4766  0.1045  99.3032   

      586     587     588       589  Pass/Fail  
0     NaN     NaN     NaN       NaN         -1  
1  0.0096  0.0201  0.0060  208.2045         -1  
2  0.0584  0.0484  0.0148   82.8602          1  
3  0.0202  0.0149  0.0044   73.8432         -1  
4  0.0202  0.0149  0.0044   73.8432         -1  

[5 rows x 592 columns]

In [4]:
# include='all'하면 입력의 모든 열이 데이터 출력으로 포함, 기본값인 'None'은 수치데이터가 있는 열만 결과에 포함
data.describe(include='all')

Time            0            1            2  \
count                  1567  1561.000000  1560.000000  1553.000000   
unique                 1534          NaN          NaN          NaN   
top     2008-10-15 01:52:00          NaN          NaN          NaN   
freq                      3          NaN          NaN          NaN   
mean                    NaN  3014.452896  2495.850231  2200.547318   
std                     NaN    73.621787    80.407705    29.513152   
min                     NaN  2743.240000  2158.750000  2060.660000   
25%                     NaN  2966.260000  2452.247500  2181.044400   
50%                     NaN  3011.490000  2499.405000  2201.066700   
75%                     NaN  3056.650000  2538.822500  2218.055500   
max                     NaN  3356.350000  2846.440000  2315.266700   

                  3            4       5            6            7  \
count   1553.000000  1553.000000  1553.0  1553.000000  1558.000000   
unique          NaN          NaN     NaN          NaN          NaN   
top             NaN          NaN     NaN          NaN          NaN   
freq            NaN          NaN     NaN          NaN          NaN   
mean    1396.376627     4.197013   100.0   101.112908     0.121822   
std      441.691640    56.355540     0.0     6.237214     0.008961   
min        0.000000     0.681500   100.0    82.131100     0.000000   
25%     1081.875800     1.017700   100.0    97.920000     0.121100   
50%     1285.214400     1.316800   100.0   101.512200     0.122400   
75%     1591.223500     1.525700   100.0   104.586700     0.123800   
max     3715.041700  1114.536600   100.0   129.252200     0.128600   

                  8  ...         581          582          583          584  \
count   1565.000000  ...  618.000000  1566.000000  1566.000000  1566.000000   
unique          NaN  ...         NaN          NaN          NaN          NaN   
top             NaN  ...         NaN          NaN          NaN          NaN   
freq            NaN  ...         NaN          NaN          NaN          NaN   
mean       1.462862  ...   97.934373     0.500096     0.015318     0.003847   
std        0.073897  ...   87.520966     0.003404     0.017180     0.003720   
min        1.191000  ...    0.000000     0.477800     0.006000     0.001700   
25%        1.411200  ...   46.184900     0.497900     0.011600     0.003100   
50%        1.461600  ...   72.288900     0.500200     0.013800     0.003600   
75%        1.516900  ...  116.539150     0.502375     0.016500     0.004100   
max        1.656400  ...  737.304800     0.509800     0.476600     0.104500   

                585          586          587          588          589  \
count   1566.000000  1566.000000  1566.000000  1566.000000  1566.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       3.067826     0.021458     0.016475     0.005283    99.670066   
std        3.578033     0.012358     0.008808     0.002867    93.891919   
min        1.197500    -0.016900     0.003200     0.001000     0.000000   
25%        2.306500     0.013425     0.010600     0.003300    44.368600   
50%        2.757650     0.020500     0.014800     0.004600    71.900500   
75%        3.295175     0.027600     0.020300     0.006400   114.749700   
max       99.303200     0.102800     0.079900     0.028600   737.304800   

          Pass/Fail  
count   1567.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean      -0.867262  
std        0.498010  
min       -1.000000  
25%       -1.000000  
50%       -1.000000  
75%       -1.000000  
max        1.000000  

[11 rows x 592 columns]

## 1. Data Preprocessing

In [5]:
# Check data's Null values
data.isnull().any().any()

True

위의 결과를 통해 데이터에 결측값이 있다는 것을 확인했고, Null값이 900개가 넘는 Column은 삭제하고 나머지 Column의 Null값은 0으로 대체하는 전처리 방법을 택한다.

In [6]:
# 각 Column별로 결측값의 수 확인
d=data.isnull().sum()
d

Time          0
0             6
1             7
2            14
3            14
             ..
586           1
587           1
588           1
589           1
Pass/Fail     0
Length: 592, dtype: int64

In [7]:
# NaN값이 900개가 넘는 경우 j배열에 저장
j = []

for i in d.keys():
    if(d[i]>900):
        print(i,d[i])
        j.append(i)

85 1341
109 1018
110 1018
111 1018
157 1429
158 1429
220 1341
244 1018
245 1018
246 1018
292 1429
293 1429
358 1341
382 1018
383 1018
384 1018
492 1341
516 1018
517 1018
518 1018
578 949
579 949
580 949
581 949


In [8]:
# Get rid of 'Time' column because it's meaningless
data=data.drop(['Time'],axis=1)

# Data에서 결측값 900개 넘는 j Column 삭제
data.drop(j, axis=1, inplace=True)
print(data.shape)
data

(1567, 567)


0        1          2          3       4      5         6       7  \
0     3030.93  2564.00  2187.7333  1411.1265  1.3602  100.0   97.6133  0.1242   
1     3095.78  2465.14  2230.4222  1463.6606  0.8294  100.0  102.3433  0.1247   
2     2932.61  2559.94  2186.4111  1698.0172  1.5102  100.0   95.4878  0.1241   
3     2988.72  2479.90  2199.0333   909.7926  1.3204  100.0  104.2367  0.1217   
4     3032.24  2502.87  2233.3667  1326.5200  1.5334  100.0  100.3967  0.1235   
...       ...      ...        ...        ...     ...    ...       ...     ...   
1562  2899.41  2464.36  2179.7333  3085.3781  1.4843  100.0   82.2467  0.1248   
1563  3052.31  2522.55  2198.5667  1124.6595  0.8763  100.0   98.4689  0.1205   
1564  2978.81  2379.78  2206.3000  1110.4967  0.8236  100.0   99.4122  0.1208   
1565  2894.92  2532.01  2177.0333  1183.7287  1.5726  100.0   98.7978  0.1213   
1566  2944.92  2450.76  2195.4444  2914.1792  1.5978  100.0   85.1011  0.1235   

           8       9  ...      577     582     583     584      585     586  \
0     1.5005  0.0162  ...  14.9509  0.5005  0.0118  0.0035   2.3630     NaN   
1     1.4966 -0.0005  ...  10.9003  0.5019  0.0223  0.0055   4.4447  0.0096   
2     1.4436  0.0041  ...   9.2721  0.4958  0.0157  0.0039   3.1745  0.0584   
3     1.4882 -0.0124  ...   8.5831  0.4990  0.0103  0.0025   2.0544  0.0202   
4     1.5031 -0.0031  ...  10.9698  0.4800  0.4766  0.1045  99.3032  0.0202   
...      ...     ...  ...      ...     ...     ...     ...      ...     ...   
1562  1.3424 -0.0045  ...  11.7256  0.4988  0.0143  0.0039   2.8669  0.0068   
1563  1.4333 -0.0061  ...  17.8379  0.4975  0.0131  0.0036   2.6238  0.0068   
1564     NaN     NaN  ...  17.7267  0.4987  0.0153  0.0041   3.0590  0.0197   
1565  1.4622 -0.0072  ...  19.2104  0.5004  0.0178  0.0038   3.5662  0.0262   
1566     NaN     NaN  ...  22.9183  0.4987  0.0181  0.0040   3.6275  0.0117   

         587     588       589  Pass/Fail  
0        NaN     NaN       NaN         -1  
1     0.0201  0.0060  208.2045         -1  
2     0.0484  0.0148   82.8602          1  
3     0.0149  0.0044   73.8432         -1  
4     0.0149  0.0044   73.8432         -1  
...      ...     ...       ...        ...  
1562  0.0138  0.0047  203.1720         -1  
1563  0.0138  0.0047  203.1720         -1  
1564  0.0086  0.0025   43.5231         -1  
1565  0.0245  0.0075   93.4941         -1  
1566  0.0162  0.0045  137.7844         -1  

[1567 rows x 567 columns]

In [9]:
# 남은 NaN값은 0으로 대체
data.replace(np.nan, 0 , inplace=True)
data

0        1          2          3       4      5         6       7  \
0     3030.93  2564.00  2187.7333  1411.1265  1.3602  100.0   97.6133  0.1242   
1     3095.78  2465.14  2230.4222  1463.6606  0.8294  100.0  102.3433  0.1247   
2     2932.61  2559.94  2186.4111  1698.0172  1.5102  100.0   95.4878  0.1241   
3     2988.72  2479.90  2199.0333   909.7926  1.3204  100.0  104.2367  0.1217   
4     3032.24  2502.87  2233.3667  1326.5200  1.5334  100.0  100.3967  0.1235   
...       ...      ...        ...        ...     ...    ...       ...     ...   
1562  2899.41  2464.36  2179.7333  3085.3781  1.4843  100.0   82.2467  0.1248   
1563  3052.31  2522.55  2198.5667  1124.6595  0.8763  100.0   98.4689  0.1205   
1564  2978.81  2379.78  2206.3000  1110.4967  0.8236  100.0   99.4122  0.1208   
1565  2894.92  2532.01  2177.0333  1183.7287  1.5726  100.0   98.7978  0.1213   
1566  2944.92  2450.76  2195.4444  2914.1792  1.5978  100.0   85.1011  0.1235   

           8       9  ...      577     582     583     584      585     586  \
0     1.5005  0.0162  ...  14.9509  0.5005  0.0118  0.0035   2.3630  0.0000   
1     1.4966 -0.0005  ...  10.9003  0.5019  0.0223  0.0055   4.4447  0.0096   
2     1.4436  0.0041  ...   9.2721  0.4958  0.0157  0.0039   3.1745  0.0584   
3     1.4882 -0.0124  ...   8.5831  0.4990  0.0103  0.0025   2.0544  0.0202   
4     1.5031 -0.0031  ...  10.9698  0.4800  0.4766  0.1045  99.3032  0.0202   
...      ...     ...  ...      ...     ...     ...     ...      ...     ...   
1562  1.3424 -0.0045  ...  11.7256  0.4988  0.0143  0.0039   2.8669  0.0068   
1563  1.4333 -0.0061  ...  17.8379  0.4975  0.0131  0.0036   2.6238  0.0068   
1564  0.0000  0.0000  ...  17.7267  0.4987  0.0153  0.0041   3.0590  0.0197   
1565  1.4622 -0.0072  ...  19.2104  0.5004  0.0178  0.0038   3.5662  0.0262   
1566  0.0000  0.0000  ...  22.9183  0.4987  0.0181  0.0040   3.6275  0.0117   

         587     588       589  Pass/Fail  
0     0.0000  0.0000    0.0000         -1  
1     0.0201  0.0060  208.2045         -1  
2     0.0484  0.0148   82.8602          1  
3     0.0149  0.0044   73.8432         -1  
4     0.0149  0.0044   73.8432         -1  
...      ...     ...       ...        ...  
1562  0.0138  0.0047  203.1720         -1  
1563  0.0138  0.0047  203.1720         -1  
1564  0.0086  0.0025   43.5231         -1  
1565  0.0245  0.0075   93.4941         -1  
1566  0.0162  0.0045  137.7844         -1  

[1567 rows x 567 columns]

In [10]:
# Check data's Null values
data.isnull().any().any()

False

### Data Scaling
- 1. StandardScaler
- 2. PCA 차원축소

In [11]:
# Train/Test Dataset split을 위해 Dataset을 Data/Target값으로 분리
X = data.iloc[:, 1:-1]
y = data.iloc[:, -1]

(1) StandardScaler: 각 feature의 평균을 0, 분산을 1로 변경, 모든 feature들이 같은 Scale을 갖게된다.

In [12]:
# 먼저 데이터를 학습용과 테스트용으로 분할
# test_test_split의 random_state를 제외한 다른 하이퍼파라미터는 default로 두기
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=50)

#### ※ Scaler를 사용하기 전에 주의할 점
scaler는 fit과 transform 매서드를 지니고 있다. fit 매서드로 데이터 변환을 학습하고, transform 매서드로 실제 데이터의 스케일을 조정한다.
이때! fit 매서드는 학습용 데이터에만 적용해야한다. 그 후 transform 매서드를 학습용 데이터와 테스트 데이터에 적용한다.

In [13]:
from sklearn.preprocessing import StandardScaler 
sc= StandardScaler()
X_sc = sc.fit_transform(X)
X_train_sc = sc.fit_transform(X_train)
print('Features Min value before Scale : \n {}'. format(X_train.min(axis=0)))
print('Features Max value before Scale : \n {}'. format(X_train.max(axis=0)))
print('Features Min value after Scale : \n {}'. format(X_train_sc.min(axis=0)))
print('Features Max value after Scale : \n {}'. format(X_train_sc.max(axis=0)))
X_test_sc = sc.transform(X_test)

Features Min value before Scale : 
 1      0.0000
2      0.0000
3      0.0000
4      0.0000
5      0.0000
        ...  
585    1.1975
586   -0.0169
587    0.0000
588    0.0000
589    0.0000
Length: 565, dtype: float64
Features Max value before Scale : 
 1      2846.4400
2      2315.2667
3      3619.7397
4      1114.5366
5       100.0000
         ...    
585      99.3032
586       0.1028
587       0.0799
588       0.0286
589     737.3048
Length: 565, dtype: float64
Features Min value after Scale : 
 [-14.93295364 -10.1911972   -2.99547149  -0.07888304 -10.28679648
  -8.61637696  -9.35095076 -15.31717516  -3.50082724  -3.80982668
 -23.5643244  -23.02456152   0.          -3.14283729 -14.5902196
 -13.87870972 -19.57651073 -19.01837981 -11.80199668 -13.10583523
  -2.39047815  -8.81241159  -4.43875953  -4.81047914  -6.70640468
  -9.80801734  -5.3265128  -17.18789415  -5.64708967  -5.48543376
  -6.82636077 -26.35443862  -6.26731362 -26.06830466 -18.84492145
 -25.79922289 -33.85474969 -33.6984

(2) PCA차원축소

- 많은 feature로 구성된 다차원 데이터 세트의 차원을 축소해 새로운 차원의 데이터 세트를 생성하는 것
-  일반적으로 차원이 증가할수록, 즉 feature가 많아질수록 예측 신뢰도가 떨어지고, 과적합(overfitting)이 발생하고, 개별 feature간의 상관관계가 높을 가능성이 있음

In [14]:
from sklearn.decomposition import PCA
# Initializing PCA and fitting
# Choose minimum number of PCA features that 85% variance is retained to avoid overfitting
pca = PCA(0.85)
X_sc_pca = pca.fit_transform(X_sc) # StandardScaler에서 학습용 데이터에 적용한 것처럼 fit과 transform 둘 다를 적용해줘야 하네~~??

In [15]:
# Transform train and test datasets
X_train_pca = pca.transform(X_train_sc)
X_test_pca = pca.transform(X_test_sc)

print('PCA차원 축소 이전 X_train:', X_train_sc.shape)
print('PCA차원 축소 이전 X_test:', X_test_sc.shape)
print('PCA차원 축소 이후 X_train:', X_train_pca.shape)
print('PCA차원 축소 이후 X_test:', X_test_pca.shape)

PCA차원 축소 이전 X_train: (1175, 565)
PCA차원 축소 이전 X_test: (392, 565)
PCA차원 축소 이후 X_train: (1175, 99)
PCA차원 축소 이후 X_test: (392, 99)


PCA차원 축소로 Feature가 565개에서 99개로 줄어 차원이 감소 됐음을 확인 할 수 있다.

## 2. Data Modeling

In [16]:
#inport libraries
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVC

### 2.1 교차검증
- 보통은 train set 으로 모델을 훈련, test set으로 모델을 검증한다.고정된 test set을 통해 모델의 성능을 검증하고 수정하는 과정을 반복하면, 결국 내가 만든 모델은 test set 에만 잘 동작하는 모델이 된다.즉, test set에 과적합(overfitting)하게 되므로, 다른 실제 데이터를 가져와 예측을 수행하면 엉망인 결과가 나와버리게 된다.
- 교차 검증은 train set을 train set + validation set으로 분리한 뒤, validation set을 사용해 검증하는 방식이다.

In [17]:
svc = SVC(C=1.0, kernel='rbf', gamma='scale')
scores = cross_val_score(svc, X_train_pca, y_train, cv=10, n_jobs=-1, scoring='accuracy') #cross_val_score(model,train,target,crossvalidation)
print('교차 검증별 정확도:',np.round(scores, 4))
print('평균 검증 정확도:', np.round(np.mean(scores), 4))

교차 검증별 정확도: [0.9322 0.9322 0.9322 0.9322 0.9322 0.9402 0.9316 0.9316 0.9316 0.9316]
평균 검증 정확도: 0.9328


### 2.2 모델 학습
#### SVM(kernalized Support Vector Machine)이란?
- 커널 서포트 벡터 머신: 입력 데이터에서 단순한 초평면 hyperplane으로 정의되지 않는 더 복잡한 모델을 만들 수 있도록 확장한 것입니다
- 결정 경계(Decision Boundary), 즉 분류를 위한 기준 선을 정의하는 모델
- SVM은 서포트 벡터(support vectors)를 사용해서 결정 경계(Decision Boundary)를 정의하고, 분류되지 않은 점을 해당 결정 경계와 비교해서 분류하게 된다.
- 분류(Classification)나 회귀 분석(Regression)에 사용이 가능하며, 특히 분류쪽에 성능 뛰어남
- 기존의 퍼셉트론은 가장 단순하고 빠른 분류 모형이지만 결정경계가 유일하게 존재하지 않는다.
- 서포트 벡터 머신(SVM)은 퍼셉트론 기반의 모형에 가장 안정적인 결정 경계를 찾기 위해 제한 조건을 추가한 모형이라고 볼 수 있다.
- 서포트 벡터 : 클래스 사이 경계에 가깝게 위치한 데이터 포인트 (결정 경계와 이들 사이의 거리가 SVC 모델의 dual_coef_에 저장된다.)

#### 커널 기법이란?
- 데이터셋에 비선형 특성을 추가하면 선형 모델을 더 강력하게 만들 수 있음
- 하지만, 어떤 특성을 추가해야 할지 알 수 없고, 특성을 많이 추가하면 연산 비용이 커진다.
- 커널 기법 : 새로운 특성을 만들지 않고 고차원 분류기를 학습시킬 수 있도록 한다. 주어진 데이터를 고차원의 특징 공간으로 사상해, 원래의 차원에선 포이지 않던 선형(초평면)이 데이터를 분류할 수 있도록 한다.
- 고차원 공간 맵핑 방법 : 가우시안 커널, RBF (Radial Basis Function) 커널

In [18]:
# 모델 학습
svc.fit(X_train_pca, y_train)
y_pred_svm=svc.predict(X_test_pca)

## 3. Data Evaluation
- TP(true positive): 1로 예측했을 때 정답도 1인 경우
- FP(false positive): 1로 예측했지만 정답은 0인 경우
- FN(false negative): 0으로 예측했지만 정답은 1인 경우
- TN(true negative): 0으로 예측했을 때 정답도 0인 경우
- recall(재현율)=TP/(TP+FN): 
  양성 positive 케이스들에서 진짜 true 양성 positive로 예측에 성공한 확률
- precision(정밀도)=TP/(TP+FP): 
  양성 positive라고 판단한 경우 중에 진짜 true 양성 positive인 확률
- F1_score=2∗(precision∗recall)/precision+recall

In [19]:
# Import Libraries
from sklearn.metrics import confusion_matrix, accuracy_score, f1_score, recall_score, precision_score

In [20]:
### print('confusion_matrix', confusion_matrix(y_pred_svm, y_test))
print('confusion_matrix', confusion_matrix(y_pred_svm, y_test))
print('accuracy score', accuracy_score(y_pred_svm, y_test))
print('f1_score', f1_score(y_test,y_pred_svm))
print('recall_score', recall_score(y_pred_svm, y_test))
print('precision_score',precision_score(y_test,y_pred_svm))

confusion_matrix [[367  25]
 [  0   0]]
accuracy score 0.9362244897959183
f1_score 0.0
recall_score 0.0
precision_score 0.0
